WordPress/Classes/ViewRelated/Sharing/ShareAppContentPresenter.swift

```diff
@@ -0,0 +1,121 @@
+        ShareAppContentServiceRemote(wordPressComRestApi: api)
+    }()
+
+    /// In-memory cache. As long as the same presenter instance is used, there's no need to re-fetch the content everytime `shareContent` is called.
+    private var cachedContent: RemoteShareAppContent? = nil
+
+    // MARK: Initialization
+
+    /// Instantiates the presenter. When the provided account is nil, the presenter will default to anonymous API.
+    init(account: WPAccount? = nil) {
+        self.api = account?.wordPressComRestV2Api ?? .anonymousApi(userAgent: WPUserAgent.wordPress(), localeKey: WordPressComRestApi.LocaleKeyV2)
+    }
+
+    // MARK: Public Methods
+
+    /// Fetches the content needed for sharing, and presents the share sheet through the provided `sender` instance.
+    ///
+    func present(for appName: ShareAppName, in sender: UIViewController, completion: (() -> Void)? = nil) {
+        if let content = cachedContent {
+            presentShareSheet(with: content, in: sender)
+            completion?()
+            return
+        }
+
+        guard !isLoading else {
+            completion?()
+            return
+        }
+
+        isLoading = true
+
+        remote.getContent(for: appName) { [weak self] result in
+            guard let self = self else { return }
+
+            switch result {
+            case .success(let content):
+                self.cachedContent = content
+                self.presentShareSheet(with: content, in: sender)
+
+            case .failure:
+                self.showFailureNotice(in: sender)
+            }
+
+            self.isLoading = false
+            completion?()
+        }
+    }
+}
+
+// MARK: - Delegate Definition
+
+protocol ShareAppContentPresenterDelegate: AnyObject {
+    /// Delegate method called everytime the presenter updates its loading state.
+    ///
+    /// - Parameter loading: The presenter's latest loading state.
+    func didUpdateLoadingState(_ loading: Bool)
+}
+
+// MARK: - Private Helpers
+
+private extension ShareAppContentPresenter {
+    /// Presents the share sheet by using `UIActivityViewController`. Contents to be shared will be constructed from the provided `content`.
+    ///
+    /// - Parameters:
+    ///   - content: The model containing information metadata for the sharing activity.
+    ///   - viewController: The view controller that will be presenting the activity.
+    func presentShareSheet(with content: RemoteShareAppContent, in viewController: UIViewController) {
+        guard let linkURL = content.linkURL() else {
+            return
+        }
+
+        let activityItems = [
+            ShareAppTextActivityItemSource(message: content.message) as Any,
+            linkURL as Any
+        ]
+
+        let activityViewController = UIActivityViewController(activityItems: activityItems, applicationActivities: nil)
+        viewController.present(activityViewController, animated: true, completion: nil)
```

Comment:
I approved too soon 😭.  I neglected to test on the iPad and went back to do that.  On the iPad the activity controller is presented as a popover.  We'll need to define a source view or source rect for the activityViewController's popoverPresentationViewController or else we'll encounter an exception: 
> Terminating app due to uncaught exception 'NSGenericException', reason: 'UIPopoverPresentationController (<UIPopoverPresentationController: 0x7ff5e70296a0>) should have a non-nil sourceView or barButtonItem set before the presentation occurs.'

Something like `activityViewController.popoverPresentationController?.sourceView = viewController.view ` might be good enough. (Probably needs to be before the call to `.present`.)

Reply:
Thank you for noticing this. I completely missed this one! It is now addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/17006/commits/a9bb473e77a0c7a120ff11fe5475eae7e5d357f6. Tested in iPad simulator and it now works as expected.